serde/msgpack.py

```diff
@@ -1,52 +1,57 @@
 """
 Serialize and Deserialize in MsgPack format.
 """
-from typing import Any, Type
+from typing import Any, Type, Dict
 
 import msgpack
 
-from .core import T, logger
+from .core import T, SerdeError
 from .de import Deserializer, from_dict, from_tuple
-from .se import Serializer, asdict, astuple
+from .se import Serializer, to_dict, to_tuple
 
 
 class MsgPackSerializer(Serializer):
     @classmethod
-    def serialize(cls, obj, named=True, use_bin_type=True, **opts) -> str:
-        asf = asdict if named else astuple
-        if "ext_dict" in opts:
-            opts.pop("ext_dict")
-            obj_bytes = msgpack.packb(asf(obj), use_bin_type=use_bin_type, **opts)
-            maybe_ext = msgpack.ExtType(int(obj._type), obj_bytes)
+    def serialize(cls, obj, use_bin_type: bool = True, ext_type_code: int = None, **opts) -> bytes:
+        if ext_type_code is not None:
+            obj_bytes = msgpack.packb(obj, use_bin_type=use_bin_type, **opts)
+            obj_or_ext = msgpack.ExtType(ext_type_code, obj_bytes)
         else:
-            maybe_ext = asf(obj)
-        return msgpack.packb(maybe_ext, use_bin_type=use_bin_type, **opts)
+            obj_or_ext = obj
+        return msgpack.packb(obj_or_ext, use_bin_type=use_bin_type, **opts)
 
 
 class MsgPackDeserializer(Deserializer):
     @classmethod
-    def deserialize(cls, s, raw=False, use_list=False, **opts):
-        unp = msgpack.unpackb(s, raw=raw, use_list=use_list, **opts)
-        logger.debug('unpack from msgpack: {unp}')
-        return unp
+    def deserialize(cls, s, raw: bool = False, use_list: bool = False, **opts):
+        return msgpack.unpackb(s, raw=raw, use_list=use_list, **opts)
 
 
-def to_msgpack(obj: Any, se=MsgPackSerializer, **opts) -> bytes:
+def to_msgpack(obj: Any, se: Serializer = MsgPackSerializer, named: bool = True, ext_dict: Dict[int, Type] = None, **opts) -> bytes:
     """
     If `ext_dict` option is specified, `obj` is encoded as a `msgpack.ExtType`
     """
-    return se.serialize(obj, **opts)
+    ext_type_code = None
+    if ext_dict is not None:
+        obj_type = type(obj)
+        ext_type_code = next((code for code, ext_type in ext_dict.items() if obj_type is ext_type), None)
```

Comment:
I know these kind of one liners are bad practice.
But I wanted both functions to accept the same `ext_dict: Dict[int, Type]` (like in the test).
And for `to_msgpack` we have to get the key (integer type code) of the dict by searching with its value (the type).
A simple `ext_dict.get(` only works for the other direction. See `from_msgpack`.

Reply:
On second thoughts I think it would be more efficient if the `from_msgpack` accepted a reversed dict. That way, the app would do the one time reversal and lookups would be efficient in both directions.

Reply:
I think you meant `to_msgpack` because `from_msgpack` already uses a simple lookup.

Yes two dicts would be more efficient/faster. Another way could be to add the `ext_type_code` directly to the `to_msgpack` function as argument. Maybe the app already knows which type code to use.

In the original code the type code was saved as `_type` attribute. In my opinion the knowledge about the name of this special attribute should not be part of pyserde.
But with the `ext_type_code` as argument one could use it like this:
```python
d = DerivedA(i=1, s="a", j=10)
to_msgpack(d, ext_type_code=d._type)
```
So basically the app has to implement the lookup itself.

Reply:
@yukinarit Thanks for the review.

I changed `to_msgpack` to use a `Dict[Type, int]` but I am not really happy with it
I guess it will confuse at least some users and also forces the application to save two dicts with basically the same information. But for know it is good enough, maybe someone else has a better idea in future.

Also keep in mind this is now a breaking change.